---
title: "DESeq2"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{DESeq2}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)

## pre-load to avoid load messages in report
library(Glimma)
library(edgeR)
library(DESeq2)
```

## Introduction

In this vignette we present the basic features of Glimma. 
Glimma is an interactive R widget for creating plots for differential expression analysis, created using the Vega and htmlwidgets frameworks. The created plots can be embedded in R Markdown, or exported as standalone HTML documents.
The data presented here is slightly modified from the [RNAseq123](https://bioconductor.org/packages/release/workflows/html/RNAseq123.html) workflow with only a single contrast has been performed for simplicity. Here we use DESeq2 to fit the model.

To begin, the DGEList object from the workflow has been included with the package as internal data. We will convert this to a DESeq data object.

```{r}
library(Glimma)
library(edgeR)
library(DESeq2)

dge <- readRDS(system.file("RNAseq123/dge.rds", package = "Glimma"))

dds <- DESeqDataSetFromMatrix(
  countData = dge$counts,
  colData = dge$samples,
  rowData = dge$genes,
  design = ~group
)
```

## MDS Plot

The multidimensional scaling (MDS) plot is frequently used to explore differences in samples. When data has been MDS transformed, the first two dimensions explain the greatest variance between samples, and the amount of variance decreases monotonically with increasing dimension.

The Glimma MDS contains two main components:

1. a plot showing two MDS dimensions, and
2. a plot of the eigenvalues of each dimension


The Glimma MDS allows different dimensions to be plotted against each other, with the proportion of variability explained by each dimension highlighted in the barplot alongside it. The interactive MDS plot can be created simply with a single argument for a DESeqDataSet object. The points in the MDS plot can have their size, colour and shape changed based on the information that is stored in the colData of the DESeqDataSet.

```{r}
glimmaMDS(dds)
```

### Interactions with the plot
In the plot above, try:

+ Scaling the points by library size (lib_size).
+ Changing the colour of points by group using the colour_by field.
+ Changing the colour scheme using to colour points using the colourscheme field.
+ Altering the shape of points by sample sequencing lane using the shape_by field.
+ Changing the dimensions plotted on the x-axis (x_axis) to dim2 and y-axis (y_axis) to dim3.
+ Saving the plots in either PNG or SVG formats using the "Save Plot" button.

### Modifications to the pot
Some helpful customisations to the plot include:

+ `glimmaMDS(dds, width=1200, height=1200)`, which will adjust the dimensions in pixels of the created widget - default width and height is 920px.
+ `glimmaMDS(dds, continuous.color=TRUE)`, which specifies that continuous colour schemes should be used - useful for when a large number of differential selections are required.
+ `glimmaMDS(dds, groups=[vector or data frame])`, which allows changing the associated sample information such as experimental groups - this information is displayed in mouseover tooltips and can be used to adjust the plot using `scale_by`, `colour_by` and `shape_by`.


## MA Plot

The MA plot is a visualisation that plots the log-fold-change between experimental groups (M) against the mean expression across all the samples (A) for each gene.

The Glimma MA plot contains two main components:

1. a plot of summary statistics across all genes that have been tested, and
2. a plot of gene expression from individual samples for a given gene

The second plot shows gene expression from the last selected sample, which can be selected from the table or directly from the summary plot.

To create the MA plot we first need to run differential expression (DE) analysis for our data using the `DESeq` function.

```{r}
dds <- DESeq(dds, quiet=TRUE)
```

The MA plot can then be created using the `dds` object that now contains fitted results.

```{r}
glimmaMA(dds)
```


### Interactions with the plot
In the plot above, try:

+ Clicking points in the summary plot or rows in the table to plot the gene expression of the selection.
    + Clicking genes in the table after selecting individual points will remove the previous selection.
+ Searching for individual genes using the search box. The search results are displayed in the table.
    + If genes are currently selected, the search box will not function.
+ Setting a maximum value for the y-axis of the expression plot using the max_y_axis field. 
    + This allows for comparison of gene expression between genes on a comparable scale.
+ Saving the currently selected genes using the Save Data dropdown. 
    + From here, you can also choose to save the entire table.
+ Saving the summary plot or expression plot in either PNG or SVG formats, using the "Save Data" dropdown.

### Modifications to the plot
Some customisations to the plot include:

+ `glimmaMA(dds, width=1200, height=1200)`, which will adjust the dimensions in pixels of the created widget
    + Default width and height is 920px.
+ `glimmaMA(dds, continuous.color=TRUE)`, which specifies that continuous colour schemes should be used
    + Useful for when a large number of differential selections are required.
+ `glimmaMA(dds, groups=[vector or data frame])`, which allows changing the associated sample information such as experimental groups
    + This information is displayed in mouseover tooltips and can be used to adjust the plot using scale_by, colour_by and shape_by.
+ `glimmaMA(dds, status.cols=c("powderblue", "seashell", "salmon")`, which customises the colours associated with the status of each gene
    + These need to be valid CSS colour strings.
+ `glimmaMA(dds, sample.cols=colours)`, which colours each sample based on the character vector of valid CSS colour strings `colours`
    + This vector needs to be of length `ncol(dge$counts)` or `ncol(counts)` if specified.

## Saving widgets

The plots created are automatically embedded into Rmarkdown reports, but having many interactive plots can significantly slow down the page. It is instead recommended to save the plots using `htmlwidgets::saveWidget` and linking to it via markdown hyperlinks.

```{r, eval = FALSE}
# creates ma-plot.html in working directory
# link to it in Rmarkdown using [MA-plot](ma-plot.html)
htmlwidgets::saveWidget(glimmaMA(dds), "ma-plot.html")
```

## Session Info

```{r}
sessionInfo()
```

